text/0000-pack-publish-buildpack.md

```diff
@@ -0,0 +1,270 @@
+    - Create and merge a PR against the `https://github.com/buildpacks/registry-index` repo using a Gitub token (i.e. all commits in that repo will be made by the same "user"). The PR will set `yanked=true` for the buildpack version described in the issue.
+    - Close the Github issue.
+
+The following behavior will execute when the configured registry is of type `git`.
+
+1. Creates a `local` branch on the local registry index cache
+1. Makes a Git commit to the local registry index with an update setting `yank=true` in the JSON payload represeting the buildpack version as defined in [RFC-0022](https://github.com/buildpacks/rfcs/blob/master/text/0022-client-side-buildpack-registry.md)
+1. Uses Git to push the commit to the registry index's `origin`.
+1. Updates the `master` branch of the local registry cache from `origin`.
+
+Options:
+
+* `--undo` - this option will execute the same flow, but set `yanked=false`
+
+```
+$ pack yank-buildpack buildpacks/sample@0.0.1
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* `type` of registry isn't accounted for in the [Buildpack URI](https://github.com/buildpacks/rfcs/blob/master/text/0037-buildpack-uris.md) definition
+* Github is a single-point-of-failure for the official registry
+* The user must have a Github account to publish to the official registry
+* The user may only publish to the official registry via a browser
+* "publish" is ambigious (we use it in `pack package-buildpack --publish` to mean pushing an OCI image to a Docker registry).
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Infer registry type
+
+Instead of configuring the registry `type`, Pack could infer the type from the URL. For example, any URL with the host `github.com` would be interpreted as type `github`. Any local file path would be interpreted as type `git`.
+
+## Direct commits to the remote index
+
+Instead of handling Github Issues on the `buildpacks/registry` repo, the `publish-buildpack` command could open a PR against the `buildpacks/registry-index`. This has the following advantages:
+
+* The operation could be performed headlessly without storing or transporting the user's Github token.
+
+This also has the following drawbacks:
+
+* The user *must* fork the `buildpacks/registry-index`
+* The contents of a PR to the `buildpacks/registry-index` repo are not intended to be human readable, which would make it difficult for users to audit before submitting.
+* This would result in commits against the index from the actual user, which requires that they DCO sign-off the commit (it's unclear if automating the sign-off is legally acceptable).
+* This differs from [rust-lang/crates.io-index](https://github.com/rust-lang/crates.io-index), which only has commits from a single user.
+
+## Using `push-buildpack` instead of `publish-buildpack`
+
+The command to release a new version could be `pack push-buildpack` to avoid ambiguity with other commands that use a `--publish` flag.
```

Comment:
Don’t think this is the way to go here.  I’d go with changing the flag in the other command first, but I also don’t think it’s enough of a problem to break compatibility.  My choice is to use `publish-buildpack` and leave the other commands with `—publish` until we have some concrete data that there’s a large amount of confusion.

Reply:
I agree with your first remark: see https://github.com/buildpacks/rfcs/pull/75#discussion_r417480933

Speaking to the second I don't see how it wouldn't immediately be confusing. Our guide/docs will likely have these two commands side by side and we'd have to explain why the `--publish` just puts it in the OCI registry but then you have to "publish" it to the Buildpacks registry.

> One the OCI registry via `pack package-buildpack --publish` and the other to the CNB registry via `pack publish-buildpack`.

Reply:
We can make it non-breaking by providing support for the prior flag as deprecated for some period of time.

Reply:
It wouldn’t be immediately confusing because, as a first approximation, no one reads the documentation.  In practice, while sitting at a command line and typing away (likely copying and pasting from somewhere else), the context of the command lends clarity to the flag.

Reply:
@jkutner Having heard my argument, can you confirm your choice here and I'll resolve.